### PR TITLE
auth: new authentication library

### PR DESCRIPTION
**Pull request by deangereaux**
Origin: Original

New authentication library for authenticating devices, key features:
   - Two auth methods, DTLS and Challenge-Response.
   - Transport independent, initial support for serial and Bluetooth.
   - Multiple authentication instances.
   - Refer to ZAUTH RFC for more details:
      https://github.com/zephyrproject-rtos/zephyr/issues/23465

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

### CODE REVIEW

File: samples/authentication/serial/auth_server/sample.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+sample:
+  description: Auth Server serial transport sample app
+  name: Auth Server Serial
+tests:
+  sample.auth_server_serial:
+    harness: bluetooth
```
Comment (deangereaux) [inline samples/authentication/serial/auth_server/sample.yaml:6]: `bluetooth` harness probably isn't correct here, but without it I get a timeout error: `INFO    - 1/1 qemu_x86                  samples/authentication/serial/auth_server/sample.auth_server_serial FAILED Timeout (qemu)`

Any guidance on how to correctly use the harness option is appreciated.

**Review by d3zd3z — DISMISSED** [review]
I'll start with a few higher-level questions that probably would have been better for the RFC review, but I seem to have missed that.

1. When using DTLS, it appears that this sets up and uses a DTLS connection to authenticate with the server. It then appears to no longer use that DTLS connection for upcoming traffic. This makes subsequent communication with this client fairly susceptible to interception. In other words, the client can easily authenticate a DTLS connection and both the server and client can know that connection is secure, but then if subsequent communication goes over another channel, that authentication was fairly pointless, since that subsequent communication can easily be bypassed. I guess in summary, I don't really see the point in authenticating with a DTLS connection but not actually using that connection for communication.
2. This applies even moreso for a challenge response. A simple challenge response can be useful to authenticate, when communication is going over an already secured channel (it is common for the client to authenticate with a server, once a TLS/DTLS connection has been established). Over an insecure channel, it suffers the same problems as above, even if the connection is authenticated, nothing stops subsequent unsecured communication from being intercepted or tampered with.

In other words, I don't see an arguable security benefit to this library as it is. Providing something to assist with establishing a DTLS connection could be useful, although we kind of already have that in the socket layer. Providing additional authentication also seems useful, but only as something that can be layered on top of an already secured channel, primarily to authenticate in the other direction.

In addition, I'd strongly suggest using an already established standard for a challenge-response, such as SCRAM that has address many of the security concerns around a more simplistic authentication.

**Comment by d3zd3z** [comment]
There is definitely some merit in a library like this for communication over a reasonably secured channel, such as a physical connection to a device, or something where the underlying protocol uses authenticated encryption (such as BT). Using DTLS this way is still a little unusual, I think it would be better to have subsequent traffic go through the DTLS session. SASL does provide an X.509 authentication method that could be used for this type of authentication, but again, that is only really meaningful if the underlying channel is already secured in some manner.

**Comment by deangereaux** [comment]
Thanks for reviewing and the feedback. 

A little history to provide some context.  This effort was stared from a requirement for a wireless lighting control system I worked on. The nodes and mobile application communicated over a secure BT link and had to be authenticated.  Since the light controls (small low cost, low power devices) needed to authenticate each other autonomously, a certificate based approach along with DTLS was ideal.  Especially since the secure hardware elements (Microchip ATECC608B) provided the ability to securely store X.509 certs and ECC keys.  Another project I worked on needed to authenticate a heart monitor chest pads, the pad output didn’t need to be encrypted.   

I do agree with you, not having the ability to encrypt subsequent messages over a link after authenticating can be of limited value.   I seriously contemplated adding encrypting subsequent messages, but wanted to limit the scope of this initial version.  However I’m happy to add this ability to make this authentication library more useful.  The authentication library design is pluggable, so adding SCRAM is very doable, I’ll investigate adding this.

If you want, we can talk more during the next API or Devel meeting, or some other time, just email me (deang@goldenbits.com).  Bottom line: I’m happy to work with the Zephyr team to make this Authentication library useful.

**Comment by d3zd3z** [comment]
I guess it kind of comes down to what the goals are. There may be some use for a naive authentication over an unencrypted channel, when that channel could be considered secure, especially in very resource-constrained devices. As long as people using this realize that there isn't really very much security gained by doing this. Using an authentication mechanism over an already secure channel (such as a BT connection) is more useful.

Overall, this does seem like a reasonable approach.

I do wonder if something like GSSAPI would be too complex for the Zephyr environment. It has the advantage of being a standard API.

File: samples/authentication/serial/auth_client/src/main.c [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+void main(void)
+{
+	struct auth_optional_param *opt_parms = NULL;
+
+	log_init();
+
+#if defined(CONFIG_AUTH_DTLS) && defined(CONFIG_AUTH_CHALLENGE_RESPONSE)
+#error Invalid authentication config, either DTLS or Challenge-Response, not both.
+#endif
```
Comment (ceolin) [inline samples/authentication/serial/auth_client/src/main.c:190]: This kind of logic should be in the build system. These options have to be mutually exclusive

File: lib/auth/auth_dtls.c [hunk 3]
```diff
@@ -0,0 +1,959 @@
...
+	mbedtls_ssl_init(&mbed_ctx->ssl);
+	mbedtls_ssl_config_init(&mbed_ctx->conf);
+	mbedtls_x509_crt_init(&mbed_ctx->cacert);
+	mbedtls_x509_crt_init(&mbed_ctx->device_cert);
+	mbedtls_pk_init(&mbed_ctx->device_private_key);
+	mbedtls_ssl_cookie_init(&mbed_ctx->cookie_ctx);
+	mbedtls_ctr_drbg_init(&mbed_ctx->ctr_drbg);
+	mbedtls_entropy_init(&mbed_ctx->entropy);
+
+	sys_rand_get(mbed_ctx->cookie, sizeof(mbed_ctx->cookie));
```
Comment (ceolin) [inline lib/auth/auth_dtls.c:159]: don't you need cryptographically secure random numbers ? See `sys_csrand_get`

File: lib/auth/auth_dtls.c [hunk 4]
```diff
@@ -0,0 +1,959 @@
...
+					   ATUH_DTLS_MAX_TIMEOUT);
+
+	/*  Force verification.  */
+	mbedtls_ssl_conf_authmode(&mbed_ctx->conf, MBEDTLS_SSL_VERIFY_REQUIRED);
+
+
+	if ((certs->device_cert_pem.priv_key == NULL) || (certs->device_cert_pem.priv_key_size == 0)) {
+		auth_free_mbedcontext(mbed_ctx);
+		LOG_ERR("Failed to get device private key");
+		return AUTH_ERROR_DTLS_INIT_FAILED;
```
Comment (ceolin) [inline lib/auth/auth_dtls.c:688]: this pattern is repetitive across this function, just log the problem and goto an error label

File: lib/auth/Kconfig [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	bool "Authentication library"
+	help
+	  This option enables the Authentication library.
+
+if AUTH_LIB
+	config AUTH_CHALLENGE_RESPONSE
+		bool "Challenge Response (Must use TINYCRYPT and TINYCRYPT_SHA256)"
+		help
+		Selects the Challenge Response authentication method.
+		Must use TINYCRYPT and TINYCRYPT_SHA256
```
Comment (ceolin) [inline lib/auth/Kconfig:16]: This dependency should be explicit.
```
depends on TINYCRYPT
depends on   TINYCRYPT_SHA256
```

File: lib/auth/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    zephyr_library_sources_ifdef(CONFIG_AUTH_CHALLENGE_RESPONSE auth_chalresp.c)
+
+    zephyr_library_sources_ifdef(CONFIG_AUTH_DTLS auth_dtls.c)
+endif()
+
+# Is this the correct way to include Mbed TLS module directory?
+if(CONFIG_MBEDTLS)
+    zephyr_include_directories($ENV{ZEPHYR_BASE}/../modules/crypto/mbedtls/include)
+    zephyr_include_directories($ENV{ZEPHYR_BASE}/../modules/crypto/mbedtls/configs)
+endif()
```
Comment (ceolin) [inline lib/auth/CMakeLists.txt:23]: nope,  first I don't think you should check on `CONFIG_MBEDTLS`. Instead of this, `AUTH_DTLS` should depends on mbedtls and here your check should be something like:\
```
if (CONFIG_AUTH_DTLS)`
zephyr_library_link_libraries_ifdef(CONFIG_MBEDTLS mbedTLS)
endif()
```

the include path should already have been set to you.

File: lib/auth/Kconfig [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+	config AUTH_DTLS
+		bool "DTLS Authentication"
+		help
+		Selects the DTLS authentication method.
+
```
Comment (ceolin) [inline lib/auth/Kconfig:22]: something here, this code depends on mbedtls doesn't ?

File: lib/auth/auth_chalresp.c [hunk 8]
```diff
@@ -0,0 +1,637 @@
...
+ * @param msgbuf      Buffer to copy message into.
+ * @param msglen      Buffer byte length.
+ *
+ * @return true if number of bytes were received, else false.
+ */
+static bool auth_server_recv_msg(struct authenticate_conn *auth_conn, uint8_t *msgbuf, size_t msglen)
+{
+	int numbytes;
+
+	while ((int)msglen > 0) {
```
Comment (ceolin) [inline lib/auth/auth_chalresp.c:324]: this is cast is wrong, if `msgelen` is smaller than `0` it means an overflow in `msgbuf`, if it is `0` you don't need to cast it to signed type. Inspect the code it does not seem  possible `msglen` be lesser than 0.

Another problem is that this cast can result in `< 0` for a big `msglen`.

File: lib/auth/auth_dtls.c [hunk 9]
```diff
@@ -0,0 +1,959 @@
...
+		break;
+	}
+
+	case 2:
+	{
+		LOG_INF(LOG_FMT,  log_strdup(basename), line, log_strdup(str));
+		break;
+	}
+
+	case 3:
```
Comment (ceolin) [inline lib/auth/auth_dtls.c:388]: use `__fallthrough` attribute in these cases.

File: lib/auth/auth_lib.c [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+}
+
+/**
+ * @see auth_lib.h
+ */
+int auth_lib_deinit(struct authenticate_conn *auth_conn)
+{
+	/* Free any resources, nothing for now, but maybe
+	 * needed in the future */
+	return AUTH_SUCCESS;
```
Comment (ceolin) [inline lib/auth/auth_lib.c:261]: don't we need to free any resource ? What happens with mbedTLS context ?

File: lib/auth/auth_lib.c [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+
+			err = auth_init_chalresp_method(auth_conn, chal_resp);
+
+			if (err) {
+				LOG_ERR("Failed to set Challege-Response param, err: %d", err);
+				return err;
+			}
+		}
+	}
+#endif
```
Comment (ceolin) [inline lib/auth/auth_lib.c:238]: these `ifdef` don't scale well, if you have N backends this code will get very complicate. You should have an API that is implemented by the backend and here you should just call the init callback.

**Comment by ceolin** [comment]
I put some comments in parts of the code, but I share @d3zd3z concerns and I think we should elaborate / discuss more the pros/cons of adopting an existent standard

File: lib/auth/CMakeLists.txt (hunk 6 above)
Comment (deangereaux) [reply lib/auth/CMakeLists.txt:23]: Thanks for the guidance here, fixed.

File: lib/auth/Kconfig (hunk 7 above)
Comment (deangereaux) [reply lib/auth/Kconfig:24]: Fixed.

File: lib/auth/auth_chalresp.c [hunk 12]
```diff
@@ -100,32 +101,67 @@ struct auth_chalresp_result {
...
-	0xBD, 0x84, 0xDC, 0x6E, 0x5C, 0x77, 0x41, 0x58, 0xE8, 0xFB, 0x1D, 0xB9, 0x95, 0x39, 0x20, 0xE4,
-	0xC5, 0x03, 0x69, 0x9D, 0xBC, 0x53, 0x08, 0x20, 0x1E, 0xF4, 0x72, 0x8E, 0x90, 0x56, 0x49, 0xA8
+struct chalresp_instance {
+	bool in_use;
+	uint8_t chalresp_key[AUTH_SHARED_KEY_LEN];
 };
 
-/* default shared key if not set */
-static uint8_t *shared_key = default_shared_key;
+static struct chalresp_instance chalresp_inst_tbl[NUM_CHALLENGE_RESP_INST];
```
Comment (deangereaux) [inline lib/auth/auth_chalresp.c:114]: Found problem here were the Challenge-Response didn't correctly handle multiple instances.  Now it does.

File: lib/auth/auth_chalresp.c [hunk 13]
```diff
@@ -100,32 +101,67 @@ struct auth_chalresp_result {
...
 
+
 /**
- * Shared key.
- * @brief  In a production system, the shared key should be stored in a
- * secure hardware store such as an ECC608A or TrustZone.
+ * Challenge-Response instance info.  Used to set shared key for each
+ * auth instance.
  */
-static uint8_t default_shared_key[AUTH_SHARED_KEY_LEN] = {
```
Comment (deangereaux) [inline lib/auth/auth_chalresp.c]: Removed default shared key out of fear users of this library would just rely on this key.

File: lib/auth/auth_dtls.c (hunk 9 above)
Comment (deangereaux) [reply lib/auth/auth_dtls.c:389]: Fixed.

File: lib/auth/auth_dtls.c [hunk 14]
```diff
@@ -622,189 +624,20 @@ static int auth_tls_entropy(void *data, unsigned char *output, size_t len,
...
 }
 
 
-
-/* ================= external/internal funcs ==================== */
-/**
- * @see auth_internal.h
- *
- */
-int auth_init_dtls_method(struct authenticate_conn *auth_conn, struct auth_dtls_certs *certs)
```
Comment (deangereaux) [inline lib/auth/auth_dtls.c]: Not sure why git thinks this entire function was removed and re-added.

File: lib/auth/auth_lib.c (hunk 11 above)
Comment (deangereaux) [reply lib/auth/auth_lib.c:203]: Makes sense, the common code should not have to know about the individual auth method details. Fixed, moved backend specific init code to the backend init function.

File: lib/auth/auth_xport_common.c [hunk 15]
```diff
@@ -1003,6 +1004,19 @@ void auth_message_hdr_to_be16(struct auth_message_frag_hdr *frag_hdr)
 	frag_hdr->payload_len = sys_cpu_to_be16(frag_hdr->payload_len);
 }
 
+/**
+ * @see auth_internal.h
+ */
+void auth_get_random(uint8_t *buf, size_t num)
```
Comment (deangereaux) [inline lib/auth/auth_xport_common.c:1010]: New common Random function.

File: lib/auth/auth_lib.c (hunk 10 above)
Comment (deangereaux) [reply lib/auth/auth_lib.c]: Yes, that's an oversight.  Added code to properly de-init the Challenge-Response and DTLS auth methods.

File: lib/auth/auth_chalresp.c (hunk 8 above)
Comment (deangereaux) [reply lib/auth/auth_chalresp.c]: Fixed.

File: lib/auth/Kconfig (hunk 5 above)
Comment (deangereaux) [reply lib/auth/Kconfig]: Fixed.

File: lib/auth/auth_dtls.c (hunk 4 above)
Comment (deangereaux) [reply lib/auth/auth_dtls.c]: Good point.  Added goto error label.

File: lib/auth/auth_dtls.c (hunk 3 above)
Comment (deangereaux) [reply lib/auth/auth_dtls.c]: Added new function `auth_get_random()` in auth_xport_common.c to handle SOCs with and without hardware based random number generators.

File: samples/authentication/serial/auth_client/src/main.c (hunk 2 above)
Comment (deangereaux) [reply samples/authentication/serial/auth_client/src/main.c:190]: This is actually an error, removed code.

**Comment by deangereaux** [comment]
@ceolin  Thanks for reviewing the PR.  Yes I agree we should discuss using a standard challenge-response.  Maybe during next Tuedays’ API meeting, if not Thurday's Devel call?  Hopefully @d3zd3z can join.

**Comment by deangereaux** [comment]
@d3zd3z  I added new functions to send/recv messages using DLTS after the initial handshake.  This provides some nice additional functionality for users.   For authentication methods besides DTLS,  a “hook” can be added enabling the user to write their own encrypt/decrypt code.  I’d like to get some user feedback and/or requests for this functionality before adding it.  

Regarding GSSAPI,  this might be something that’s layered on top of Zauth eventually.   One of my goals was to keep the interface simple to encourage use vs. a heavier API like GSSAPI.

File: lib/auth/auth_dtls.c (hunk 3 above)
Comment (d3zd3z) [reply lib/auth/auth_dtls.c]: In general, this code probably shouldn't be supported on SOCs that don't have hardware based random numbers. I wouldn't go out of the way to make it work, since the end result won't be secure.

File: lib/auth/auth_xport_common.c (hunk 15 above)
Comment (d3zd3z) [reply lib/auth/auth_xport_common.c:1017]: If we did agree to create a new random function, I think it would be appropriate to add it as its own change. However, I would argue that if there isn't a cryptographically secure random source available, the security added by this code is really more of an illusion. Non cryptographic random sources are rather predictable, and generally result in catastrophic weaknesses when used with cryptograph protocols, such as this.

Comment (dleach02) [reply lib/auth/auth_xport_common.c:1017]: I don't think you should be defining yet another wrapper around rand. Decide whether your subsystem requires a cryptographically secure random values. If it does then just call sys_csrand_get() directly because that shows the reader the intent. If it doesn't then call sys_rand_get(). This way, if the underlying HW doesn't support a cryptographically secure random and someone tries to enable code that requires it, a build error will occur.

**Review by dleach02 — CHANGES_REQUESTED** [review]
You need to clarify the subsystem requirements for rand()

**Comment by deangereaux** [comment]
@dleach02  Latest commit removes wrapper around  *_rand() and explicitly uses sys_csrand_get()

**Comment by deangereaux** [comment]
@ceolin @d3zd3z   I added an explicit warning in the docs about how the Challenge-Response does not provide security for subsequent messages and how DTLS can be used for follow-on secure messages.

I spent some time reviewing/read specs for GSSAPI.   It seems GSSAPI is suited for handling authentication for web base apps or enterprise level networks vs. small embedded devices which ZAUTH is targeting.  Regarding SCRAM, mentioned in this PR, might be good to add later.  All of the Challenge-Response uses cases I’ve worked on are straight forward and do not require something like SCRAM.  For example, authenticating a printer cartridge or an attached display.  But yes this all depends on the application and risk associated with exposing printer commands or displaying incorrect messages.   

Regardless, I’m happy to talk and decide what is the best path forward for this PR.
Thanks

**Comment by d3zd3z** [comment]
I do think there are decent uses of a simplistic authentication library for Zephyr, as long as we are clear that there are security tradeoffs being made with regards to code size and complexity. For something like authenticating an attached piece of hardware, and the likes, just authenticating, and then not worrying about the security of the resulting transmission probably makes sense.

**Comment by deangereaux** [comment]
@dleach02    I had to add back the insecure random number generator, `sys_rand_get()`, in order to pass CI tests. Maybe there’s a way to use `sys_csrand_get()` when running twister tests, but dang I couldn’t find it.  I did add warnings in the code,  I’ll also update the documentation warning about using the insecure generator.

**Comment by d3zd3z** [comment]
> @dleach02 I had to add back the insecure random number generator, sys_rand_get(), in order to pass CI tests.

This is a bit deeper of a problem than just this change. At one point, we had an insecure source of entropy available, but this is a pretty bad idea if it makes it into an actual device. (e.g. [Netscape](http://people.eecs.berkeley.edu/~daw/papers/ddj-netscape.html)).

@dleach02 How do we solve this for other users of the crypto random sources? Do we need to come up with a more general solution here? It is probably ok to have insecure random numbers on CI targets, but we really need to make sure that never makes it onto an actual device.

I don't think the solution is to put the weakness into this auth code. The auth code needs strong random numbers, and so should be asking for them. We just need to figure out how to test this in CI.

**Comment by deangereaux** [comment]
@d3zd3z  I look hard at the test tools and system and didn’t find any way to mock out the sys_csrand_get() call.   As a solution I added a sub function in the test code here:  test/lib/authentication/sr/main.c.

File: tests/lib/authentication/src/main.c [hunk 16]
```diff
@@ -83,3 +83,20 @@ void test_main(void)
...
+/**
+ * Stub function to mock sys_csrand_get() function which is not available
+ * when running tests.  Resolves link error.
+ *
+ * @param dst   Buffer destination for bytes.
+ * @param len   Number of bytes.
+ *
+ * @return  0 on success.
+ */
+int z_impl_sys_csrand_get(void *dst, size_t len)
```
Comment (deangereaux) [inline tests/lib/authentication/src/main.c:96]: @d3zd3z   Here's the stub function just for CI testing.

File: samples/authentication/bluetooth/central_auth/sample.yaml [hunk 17]
```diff
@@ -3,6 +3,7 @@ sample:
   name: Auth Central
 tests:
   sample.central_auth:
+    build_only: true
     harness: bluetooth
-    platform_allow:  qemu_x86
+    platform_allow: nrf52840dk_nrf52840
```
Comment (deangereaux) [inline samples/authentication/bluetooth/central_auth/sample.yaml:8]: For CI, I added the nrf52840dk_nrf52840 board as the only allowed platform.  This is successful if you run twister with --all option like this:  `twister --all samples/authentication/bluetooth/central_auth/sample.central_auth`

File: tests/lib/authentication/src/main.c [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+{
+    /* Not random, for testing purposes only. */
+    memset(dst, 42u, len);
```
Comment (d3zd3z) [inline tests/lib/authentication/src/main.c:99]: 42 is a reasonably good choice, although https://xkcd.com/221/ 4 is also common :-)

**Comment by deangereaux** [comment]
There’s only one compliance check that is failing, a typedef (in auth_xport.h,  `typedef void *auth_xport_hdl_t`) used to define an intentionally opaque transport handle.  This should be OK per the Linux coding standards

**Comment by deangereaux** [comment]
@carlescufi @d3zd3z @dleach02 @galak @MaureenHelm @nashif @ceolin @frkv   This ZAUTH PR has been aging and like a good wine it is getting better and better!!   I’d like to push this PR to the finish line here and merge.  This library will be marked as EXPERIMENTAL, as such any issues can be readily addressed reducing any risk. 
Can you help me by reviewing and approving this PR?  
Thanks, Dean

**Comment by deangereaux** [comment]
There doesn’t seem to be any interest in going forward and merging this ZAUTH PR.   Maybe my perception is incorrect, however I’m not seeing any feedback or comments.  I do think ZAUTH is useful and provides a solution for authentication use cases I personally have had to write code to solve.  However, if the larger Zephyr community is not interested in ZAUTH, then let’s close this PR and put it on the shelf.

**Comment by deangereaux** [comment]
Closing.
